packages/flutter/lib/src/widgets/focus_traversal.dart

```diff
@@ -560,115 +733,115 @@ mixin DirectionalFocusTraversalPolicyMixin on FocusTraversalPolicy {
-    // through scopes until an ultimately focusable item is found, a scope
-    // doesn't have a focusedChild, or a non-scope is encountered.
-    FocusNode candidate = scope.focusedChild;
-    if (candidate == null) {
-      if (scope.traversalChildren.isNotEmpty) {
-        candidate = scope.traversalChildren.first;
-      } else {
-        candidate = currentNode;
-      }
+  /// Finds the common Directional ancestor of an entire list of groups.
+  static TextDirection commonDirectionalityOf(List<_ReadingOrderSortData> list) {
+    final List<List<Directionality>> allAncestors = list.map<List<Directionality>>((_ReadingOrderSortData member) => member.directionalAncestors.toList()).toList();
+    Set<Directionality> common;
+    for (final List<Directionality> ancestorList in allAncestors) {
+      final Set<Directionality> ancestorSet = ancestorList.toSet();
+      common ??= ancestorSet;
+      common = common.intersection(ancestorSet);
     }
-    while (candidate is FocusScopeNode && candidate.focusedChild != null) {
-      final FocusScopeNode candidateScope = candidate as FocusScopeNode;
-      candidate = candidateScope.focusedChild;
+    if (common.isEmpty) {
+      // If there is no common ancestor, then arbitrarily pick the
+      // directionality of the first group, which is the equivalent of the "first
+      // strongly typed" item in a bidi algorithm.
+      return list.first.directionality;
     }
-    return candidate;
+    // Find the closest common ancestor. The memberAncestors list contains the
+    // ancestors for all members, but the first member's ancestry was
+    // added in order from nearest to furthest, so we can still use that
+    // to determine the closest one.
+    return list.first.directionalAncestors.firstWhere(common.contains).textDirection;
   }
 
-  // Moves the focus to the next or previous node, depending on whether forward
-  // is true or not.
-  bool _move(FocusNode currentNode, {@required bool forward}) {
-    if (currentNode == null) {
-      return false;
-    }
-    final FocusScopeNode nearestScope = currentNode.nearestScope;
-    invalidateScopeData(nearestScope);
-    final FocusNode focusedChild = nearestScope.focusedChild;
-    if (focusedChild == null) {
-      final FocusNode firstFocus = findFirstFocus(currentNode);
-      if (firstFocus != null) {
-        _focusAndEnsureVisible(
-            firstFocus,
-            alignmentPolicy: forward
-                ? ScrollPositionAlignmentPolicy.keepVisibleAtEnd
-                : ScrollPositionAlignmentPolicy.keepVisibleAtStart,
-        );
+  /// Returns the list of Directionality ancestors, in order from nearest to
+  /// furthest.
+  Iterable<Directionality> get directionalAncestors {
+    List<Directionality> getDirectionalityAncestors(BuildContext context) {
+      final List<Directionality> result = <Directionality>[];
+      context.visitAncestorElements((Element element) {
```

Comment:
Done.  I couldn't use `Directionality.of`, since I actually need to find the common ancestor widget, not the next `TextDirection` that matches, but I was able to make my own traversal using `getElementForInheritedWidgetOfExactType` and a concocted `getParent` for the element.